csharp/src/Ice/TraceUtil.cs

```diff
@@ -2,315 +2,190 @@
 
-                    case 2:
-                        {
-                            s.Write("(compressed; compress response, if any)");
-                            break;
-                        }
+        internal static void TraceSend(Communicator communicator, Span<byte> header) =>
+            Trace("sending ", communicator, header);
 
-                    default:
-                        {
-                            s.Write("(unknown)");
-                            break;
-                        }
-                }
+        internal static void TraceReceived(Communicator communicator, Span<byte> header) =>
+            Trace("received ", communicator, header);
 
-                int size = str.ReadInt();
-                s.Write("\nmessage size = " + size);
-                return type;
-            }
-            catch (System.IO.IOException)
+        internal static void Trace(string heading, Communicator communicator, Span<byte> header)
+        {
+            if (communicator.TraceLevels.Protocol >= 1)
             {
-                Debug.Assert(false);
-                return 0;
+                using var s = new StringWriter(CultureInfo.CurrentCulture);
+                s.Write(heading);
+                s.Write(GetFrameTypeAsString((Ice1Definitions.FrameType)header[8]));
+                PrintHeader(header, s);
+                communicator.Logger.Trace(communicator.TraceLevels.ProtocolCat, s.ToString());
             }
         }
 
-        private static Ice1Definitions.FrameType PrintMessage(System.IO.StringWriter s, InputStream str)
+        private static void PrintHeader(Span<byte> header, StringWriter s)
         {
-            var type = (Ice1Definitions.FrameType) PrintHeader(s, str);
-
-            switch (type)
-            {
-                case Ice1Definitions.FrameType.CloseConnection:
-                case Ice1Definitions.FrameType.ValidateConnection:
-                    {
-                        // We're done.
-                        break;
-                    }
-
-                case Ice1Definitions.FrameType.Request:
-                    {
-                        PrintRequest(s, str);
-                        break;
-                    }
-
-                case Ice1Definitions.FrameType.RequestBatch:
-                    {
-                        PrintBatchRequest(s, str);
-                        break;
-                    }
-
-                case Ice1Definitions.FrameType.Reply:
-                    {
-                        PrintReply(s, str);
-                        break;
-                    }
-
-                default:
-                    {
-                        s.Write("(unknown)");
-                        break;
-                    }
-            }
-
-            return type;
+            byte messageType = header[8];
+            s.Write("\nmessage type = ");
+            s.Write(messageType);
+            s.Write((Ice1Definitions.FrameType)messageType switch
+            {
```

Comment:
This is redundant with the code for the GetFrameTypeAsString method at the end of this file.

Reply:
Fixed